pkg/ccl/changefeedccl/event_processing.go

```diff
@@ -240,3 +330,238 @@ func (c *kvEventToRowConsumer) ConsumeEvent(ctx context.Context, ev kvevent.Even
+	consumers := make([]eventConsumer, c.numWorkers)
+	for i := int64(0); i < c.numWorkers; i++ {
+		consumer, err := c.makeConsumer()
+		if err != nil {
+			return err
+		}
+		consumers[i] = consumer
+	}
+
+	for i := int64(0); i < c.numWorkers; i++ {
+		c.workerCh[i] = make(chan kvevent.Event, c.workerChSize)
+
+		// c.g.GoCtx requires a func(context.Context), so
+		// a closure is used to pass the consumer and id to
+		// the worker.
+		id := i
+		consumer := consumers[i]
+		workerClosure := func(ctx2 context.Context) error {
+			return c.workerLoop(ctx2, consumer, id)
+		}
+		c.g.GoCtx(workerClosure)
+	}
+	return nil
+}
+
+func (c *parallelEventConsumer) workerLoop(
+	ctx context.Context, consumer eventConsumer, id int64,
+) error {
+	log.Infof(ctx, "started parallel event consumer worker %d", id)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-c.doneCh:
+			return nil
+		case e := <-c.workerCh[id]:
+			if err := consumer.ConsumeEvent(ctx, e); err != nil {
+				return c.setWorkerError(err)
+			}
+			c.decInFlight()
+		}
+	}
+}
+
+func (c *parallelEventConsumer) incInFlight() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.mu.inFlight++
+	c.metrics.ParallelConsumerInFlightEvents.Inc(1)
+}
+
+func (c *parallelEventConsumer) decInFlight() {
+	c.mu.Lock()
+	c.mu.inFlight--
+	c.metrics.ParallelConsumerInFlightEvents.Dec(1)
+	send := c.mu.waiting && c.mu.inFlight == 0
+	c.mu.Unlock()
+
+	// If someone is waiting on a flush, signal to them
+	// that there are no more events.
+	if send {
+		c.flushCh <- struct{}{}
+	}
+}
+
+func (c *parallelEventConsumer) setWorkerError(err error) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.mu.termErr == nil {
+		c.mu.termErr = err
+		close(c.termCh)
+	}
+
+	return err
+}
+
+// Flush flushes the consumer by blocking until all events are consumed,
+// or until there is an error.
+func (c *parallelEventConsumer) Flush(ctx context.Context) error {
+	log.Info(ctx, "flushing parallel event consumer")
+
+	startTime := timeutil.Now().UnixNano()
+	defer func() {
+		time := timeutil.Now().UnixNano()
+		c.metrics.ParallelConsumerFlushNanos.Inc(time - startTime)
+	}()
+
+	needFlush := func() bool {
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		if c.mu.inFlight > 0 {
+			c.mu.waiting = true
+		}
+		return c.mu.waiting
+	}
+
+	if !needFlush() {
+		return nil
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-c.termCh:
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		return c.mu.termErr
+	case <-c.flushCh:
+		c.mu.Lock()
+		c.mu.waiting = false
+		c.mu.flushFrontier = c.spanFrontier.Frontier()
+		c.mu.Unlock()
+		return nil
+	}
+}
+
+func (c *parallelEventConsumer) Frontier() hlc.Timestamp {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.mu.flushFrontier
+}
+
+func (c *parallelEventConsumer) Close() error {
+	// If there is a terminal error, the parallel consumer is shutting down.
+	c.mu.Lock()
+	if c.mu.termErr != nil {
+		c.mu.Unlock()
+		return c.mu.termErr
+	}
+	c.mu.Unlock()
```

Comment:
if it's shutting down, do we still need above? I think we can just close doneCh and wait.

Reply:
Yeah you're absolutely right. The worker that saw the error is guaranteed to return the error it saw, so it that error will be returned by `Wait`.